### PR TITLE
git-server: add support for project alias

### PR DESCRIPTION
**Pull request by xphoniex**
Closes https://github.com/radicle-dev/radicle-client-services/issues/72

### CODE REVIEW

**Comment by xphoniex** [comment]
re listening for new projects, I think we can standardize the alias format (e.g. ending in ".git") and re-fill our mapping on cache misses. We don't expect to get too many misses anyway, and standardizing format helps us only focus on cases that matter, so we should get good performance.

I can add this in, if it's a go.

File: git-server/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ tokio = { version = "1.2", features = ["macros", "rt", "rt-multi-thread", "sync"
 tracing = "0.1"
 tracing-subscriber = "0.2"
 warp = { version = "0.3.1", features = ["tls"] }
+radicle-source = { version = "0.2.0", features = ["syntax"] }
```
Comment (cloudhead) [inline git-server/Cargo.toml:31]: There's no need for the `syntax` feature here, it handles syntax highlighting :)

File: git-server/src/lib.rs [hunk 2]
```diff
@@ -112,6 +113,10 @@ impl Context {
 pub async fn run(options: Options) {
     let ctx = Context::try_from(&options).expect("Failed to create context from service options");
 
+    let mut aliases = HashMap::new();
+    grab_aliases(&options, &mut aliases);
+    let aliases = Arc::new(aliases);
```
Comment (cloudhead) [inline git-server/src/lib.rs:118]: Aliases should go inside `Context`, then `grab_aliases` can be a method on `Context`.

File: git-server/src/lib.rs [hunk 3]
```diff
@@ -385,3 +393,41 @@ fn gen_random_string() -> String {
...
+    let mut duplicates: Vec<String> = Vec::new();
+    let identities = identities::any::list(&storage)
+        .expect("Failed to list identities in storage");
+    for identity in identities {
+        match identity {
+            Ok(Project(project)) => {
+                let urn = Namespace::try_from(project.urn().encode_id().as_str())
+                    .unwrap().to_string();
+                let name = (&project.payload().subject.name).to_string() + ".git";
+                tracing::info!("alias \"{}\" for \"{}\"", name, urn);
```
Comment (cloudhead) [inline git-server/src/lib.rs:418]: If you use `"{:?}"` on a string, it will be quoted.

File: git-server/src/lib.rs (hunk 2 above)
Comment (cloudhead) [reply git-server/src/lib.rs:118]: We should also use `Urn` as the value to make use of the type checker.

**Comment by cloudhead** [comment]
> re listening for new projects, I think we can standardize the alias format (e.g. ending in ".git") and re-fill our mapping on cache misses. We don't expect to get too many misses anyway, and standardizing format helps us only focus on cases that matter, so we should get good performance.
> 
> I can add this in, if it's a go.

Hmm, the thing is I'd like `.git` to be allowed for project IDs but required for aliases. So how about this:

```
if path ends with ".git" {
  if path is a valid project id {
     lookup project
  } else {
    lookup alias
  }
} else {
  lookup project
}
```

File: git-server/Cargo.toml (hunk 1 above)
Comment (xphoniex) [reply git-server/Cargo.toml]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/files#diff-7150a884e7aeebff16095bb654e6c671a89cbb34d28864e467368c53ea3bff87R31

File: git-server/src/lib.rs (hunk 2 above)
Comment (xphoniex) [reply git-server/src/lib.rs]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/commits/8e795c8bba5f1e1edb21d3dd531c6a8bea3d75ff#diff-b672a15ad14f54528cebd77aee76d3007962d44bda987b3e0a34597ec80fd6b1R48

File: git-server/src/lib.rs (hunk 3 above)
Comment (xphoniex) [reply git-server/src/lib.rs]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/commits/8e795c8bba5f1e1edb21d3dd531c6a8bea3d75ff#diff-b672a15ad14f54528cebd77aee76d3007962d44bda987b3e0a34597ec80fd6b1R134

File: git-server/src/lib.rs [hunk 4]
```diff
@@ -385,3 +393,41 @@ fn gen_random_string() -> String {
...
+                }
+                map.insert(name, urn);
+            },
+            _ => {},
+        }
+    }
+
+    for duplicate in duplicates {
+        map.remove(&duplicate);
+    }
```
[the remark this answers is not in the document]
Comment (xphoniex) [reply git-server/src/lib.rs]: From original [issue](https://github.com/radicle-dev/radicle-client-services/issues/72):
> Note that project names are not unique, so it's possible for there to be duplicates. This should be handled somehow, perhaps by not supporting aliases for projects with non-unique names.

**Comment by xphoniex** [comment]
> Hmm, the thing is I'd like `.git` to be allowed for project IDs but required for aliases. So how about this:
> 
> ```
> if path ends with ".git" {
>   if path is a valid project id {
>     lookup project
>   } else {
>     lookup alias
>   }
> } else {
>   lookup project
> }
> ```
> 
> `<project-id>` -> OK `<project-id>.git` -> OK `<alias>.git` -> -> OK `<alias>` -> ERROR
> 
> The only case where this fails if you use another project's id as your project name, which I don't see why anyone would do laughing

This should be okay now with https://github.com/radicle-dev/radicle-client-services/pull/74/commits/3f5776a431cd2377d9749cc333a8196110d47858. 

And if the remote is set to `http://localhost:8778/hnrkyfoxs5xp45wg1p8q3dkh74qyebqaeunqo.git` we just redirect it to `hnrkyfoxs5xp45wg1p8q3dkh74qyebqaeunqo`.

Though, I noticed non-existant aliases, trigger 3 updates:
```
Nov 30 12:54:49.807  INFO radicle_git_server: version 0.1.0-8e795c8
Nov 30 12:54:49.923  INFO radicle_git_server: alias "delete-this.git" for "hnrkyfoxs5xp45wg1p8q3dkh74qyebqaeunqo"
Nov 30 12:54:50.084  INFO radicle_git_server: git version 2.28.0
Nov 30 12:54:50.088  INFO Server::run{addr=0.0.0.0:8778}: warp::server: listening on http://0.0.0.0:8778
doesnt exist... rebuilding cache
Nov 30 12:54:51.583  INFO radicle_git_server: alias "delete-this.git" for "hnrkyfoxs5xp45wg1p8q3dkh74qyebqaeunqo"
rebuilt cache
Nov 30 12:54:51.585 ERROR radicle_git_server: unauthorized
Nov 30 12:54:51.586  INFO radicle_git_server: 127.0.0.1:50776 "GET /delete-this2.git/info/refs HTTP/1.1" 401 "-" "git/2.28.0" 6.724034ms
doesnt exist... rebuilding cache
Nov 30 12:54:53.091  INFO radicle_git_server: alias "delete-this.git" for "hnrkyfoxs5xp45wg1p8q3dkh74qyebqaeunqo"
rebuilt cache
Nov 30 12:54:53.415  INFO radicle_git_server: 127.0.0.1:50776 "GET /delete-this2.git/info/refs HTTP/1.1" 200 "-" "git/2.28.0" 340.153381ms
doesnt exist... rebuilding cache
Nov 30 12:54:53.768  INFO radicle_git_server: alias "delete-this.git" for "hnrkyfoxs5xp45wg1p8q3dkh74qyebqaeunqo"
rebuilt cache
Nov 30 12:54:54.267  INFO radicle_git_server: 127.0.0.1:50776 "POST /delete-this2.git/git-receive-pack HTTP/1.1" 200 "-" "git/2.28.0" 513.537247ms
```

Twice on `/info/refs`, once before authentication and once after. And on `/git-receive-pack`. This should however be ok for now.

File: git-server/src/lib.rs [hunk 5]
```diff
@@ -182,11 +229,30 @@ async fn git_handler(
     headers: HeaderMap,
     body: impl Buf,
     remote: Option<net::SocketAddr>,
-    namespace: String,
+    mut namespace: String,
     request: warp::filters::path::Tail,
     query: String,
 ) -> Result<Box<dyn Reply>, Rejection> {
     let remote = remote.expect("There is always a remote for HTTP connections");
+    if namespace[namespace.len() - 4..].eq(".git") {
```
Comment (cloudhead) [inline git-server/src/lib.rs:237]: Perhaps you can use `String::ends_with` here

File: git-server/src/lib.rs [hunk 6]
```diff
@@ -182,11 +229,30 @@ async fn git_handler(
...
+        let ns_id = &namespace[..namespace.len() - 4];
```
Comment (cloudhead) [inline git-server/src/lib.rs:238]: And `strip_suffix(".git")` here

File: git-server/src/lib.rs [hunk 7]
```diff
@@ -106,11 +112,52 @@ impl Context {
 
         Ok(())
     }
+
+    /// Grabs projects names and uses projects' names as alias for urns
+    fn grab_aliases(&self) -> Result<(), Error> {
```
Comment (cloudhead) [inline git-server/src/lib.rs:117]: Can we call this something more descriptive, like `populate_aliases` or `load_aliases`?

File: git-server/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cloudhead) [reply git-server/src/lib.rs]: Ah I see. I think we're maybe doing too much for the user here. How about we don't clear the cache everytime, and if there's a duplicate, we just don't overwrite the alias in the cache and log a warning to the user.

[the remark this answers is not in the document]
Comment (cloudhead) [reply git-server/src/lib.rs]: Something like:
```
if map.contains(name) {
   warn!("duplicate project name {}", name)
} else {
   map.insert(name, urn)
}
```

File: git-server/src/lib.rs (hunk 7 above)
Comment (xphoniex) [reply git-server/src/lib.rs]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/commits/67c4936ad316ef7490d2c15d4b88ac055b09ffe9#diff-b672a15ad14f54528cebd77aee76d3007962d44bda987b3e0a34597ec80fd6b1R116-R117

File: git-server/src/lib.rs (hunk 6 above)
Comment (xphoniex) [reply git-server/src/lib.rs]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/commits/67c4936ad316ef7490d2c15d4b88ac055b09ffe9#diff-b672a15ad14f54528cebd77aee76d3007962d44bda987b3e0a34597ec80fd6b1R229

File: git-server/src/lib.rs (hunk 5 above)
Comment (xphoniex) [reply git-server/src/lib.rs]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/commits/67c4936ad316ef7490d2c15d4b88ac055b09ffe9#diff-b672a15ad14f54528cebd77aee76d3007962d44bda987b3e0a34597ec80fd6b1R228

File: git-server/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xphoniex) [reply git-server/src/lib.rs]: Fixed https://github.com/radicle-dev/radicle-client-services/pull/74/commits/67c4936ad316ef7490d2c15d4b88ac055b09ffe9#diff-b672a15ad14f54528cebd77aee76d3007962d44bda987b3e0a34597ec80fd6b1L139-R137

**Review by cloudhead — APPROVED** [review]
Nice work!

**Comment by cloudhead** [comment]
run `cargo clippy --all` to fix the lint

**Pull request merged**